src/concrete/ml/quantization/quantized_ops.py

```diff
@@ -217,20 +222,113 @@ def q_impl(
+        p = input2_q_values.shape[-2]
+
+        # FIXME: https://github.com/zama-ai/concrete-internal/issues/512
+        def enc_mul(x, y):
+            """Encrypted multiplication.
+
+            Args:
+                x (numpy.ndarray): first input numpy array
+                y (numpy.ndarray): second input numpy array
+
+            Returns:
+                numpy.ndarray: result of the multiplication
+            """
+            with tag("pbs_multiplication"):
+                add = x + y
+                sub = x - y
+                with tag(self.op_instance_name + ".pbs_matmul_rounding_add"):
+                    # Apply Concrete rounding (if relevant)
+                    add = self.cnp_round(add, calibrate_rounding, key="add")
+                with tag(self.op_instance_name + ".pbs_matmul_rounding_sub"):
+                    # Apply Concrete rounding (if relevant)
+                    sub = self.cnp_round(sub, calibrate_rounding, key="sub")
+
+                add_pow = (add.astype(numpy.float64)) ** 2
+                sub_pow = (sub.astype(numpy.float64)) ** 2
+                add_pow_divide = (add_pow / 4.0).astype(numpy.int64)
+                sub_pow_divide = (sub_pow / 4.0).astype(numpy.int64)
+
+            return add_pow_divide - sub_pow_divide
+
+        # FIXME: https://github.com/zama-ai/concrete-internal/issues/512
+        def matmul(a, b):
+            with tag("Encrypted MatMul"):
+                # Determine whether inputs are 2D or 3D
+                a_3d = a.ndim == 3
+                b_3d = b.ndim == 3
+
+                if a_3d:
+                    batch_a, m, n = a.shape
+                else:
+                    m, n = a.shape
+                    batch_a = 1
+
+                if b_3d:
+                    batch_b, n_b, p = b.shape
+                else:
+                    n_b, p = b.shape
+                    batch_b = 1
+
+                assert_true(n == n_b, "Inner dimensions do not match for matrix multiplication")
+                assert (
+                    batch_a == batch_b or batch_a == 1 or batch_b == 1
+                ), "Batch sizes must be equal or one must be 1"
+
+                batch_size = batch_a
+                c = zeros(shape=(batch_size, m, p))
+
+                for i in range(batch_size):
+                    a_slice = a[i] if a_3d else a
+                    b_slice = b[i] if b_3d else b
+
+                    # Reshape for element-wise multiplication
+                    a_reshaped = a_slice.reshape((m, n, 1))
+                    b_reshaped = b_slice.reshape((1, n, p))
+                    enc_mul_result = enc_mul(a_reshaped, b_reshaped)
+                    c[i] = numpy.sum(enc_mul_result, axis=1)
+
+                # If both inputs were 2D, remove the first dimension
+                if not a_3d and not b_3d:
+                    c = numpy.squeeze(c, axis=0)
+
+                return c
+
+        # FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/4127
```

Comment:
same please

Reply:
Done

Reply:
still don't see a comment here @jfrery 😅

Reply:
Should be good now

Reply:
thanks !